src/core/lifetime.d

```diff
@@ -2108,10 +2108,67 @@ private T trustedMoveImpl(T)(return scope ref T source) @trusted
     move(x, x);
 }
 
+// @@@REVIEW: make this public core.internal.traits?
+private enum bool hasContextPointers(T) = {
+    static if (__traits(isStaticArray, T))
+        return .hasContextPointers!(typeof(T.init[0]));
+    else static if (is(T == struct))
+    {
+        import core.internal.traits : anySatisfy;
+        return __traits(isNested, T) || anySatisfy!(.hasContextPointers, typeof(T.tupleof));
+    }
+    else return false;
+} ();
+
+@safe @nogc nothrow pure unittest
+{
+    static assert(!hasContextPointers!int);
+    static assert(!hasContextPointers!(void*));
+
+    static struct S {}
+    static assert(!hasContextPointers!S);
+    static assert(!hasContextPointers!(S[1]));
+
+    struct Nested
+    {
+        void foo() {}
+    }
+
+    static assert(hasContextPointers!Nested);
+    static assert(hasContextPointers!(Nested[1]));
+
+    static struct OneLevel
+    {
+        int before;
+        Nested n;
+        int after;
+    }
+
+    static assert(hasContextPointers!OneLevel);
+    static assert(hasContextPointers!(OneLevel[1]));
+
+    static struct TwoLevels
+    {
+        int before;
+        OneLevel o;
+        int after;
+    }
+
+    static assert(hasContextPointers!TwoLevels);
+    static assert(hasContextPointers!(TwoLevels[1]));
+
+    union U
+    {
+        Nested n;
+    }
+
+    static assert(!hasContextPointers!U);
```

Comment:
Hmm, I don't know exactly what the mechanics for context pointers and unions are, but since this code is working:

```d
auto fun()
{
    int a;
    struct S { int f() {return a;} }
    
    union T
    {
         S b;   
    }
    
    T t;
    return t;
}

void main()
{
    auto s = fun();
    int t = s.b.f();
}
```

Doesn't that mean that `T` has a context pointer?

Reply:
Unions can yield false positives. They're not discriminated. The danger of wiping context pointers is destructors, and unions don't have those. It's the caller's job to ascertain discrimination, including safe default discriminator in .init.

If ever needed, we could add a `mayHaveContextPointers` that would include unions and static arrays of void.

Reply:
Your code actually raises a question to the [spec](https://dlang.org/spec/struct.html#default_union_init):
> Unions are by default initialized to whatever the Initializer for the first field is, and if none is supplied, to the **default initializer** for the first field's type

You're not supplying initializer, so default one should be used. But default initializer would not have any context pointer set (i.e. `S.init.tupleof[$-1] is null` - which is the cause of this bug in the first place), yet in the above code it is being set. Rules for this particular case need to be specified explicitly.